Mark cache directories as purgeable

szewai opened the pull request:
#### d2e7f67539216d096f372d7ed9ef740248e68253
<pre>
Mark cache directories as purgeable
<a href="https://bugs.webkit.org/show_bug.cgi?id=247828">https://bugs.webkit.org/show_bug.cgi?id=247828</a>
rdar://101927216

Reviewed by Chris Dumez.

These directories were purgeable by CacheDelete because they were under Caches/ directory by default. After 255271@main,
the directories of a WebsiteDataStore created with UUID can be outside of Caches/ directory, so we need to explicitly
mark them purgeable to keep the old behavior. Notice the files are unmarked if they are modified after being marked, so
we set the flag before process exits or session is destroyed.

* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::setExcludedFromBackup):
(WTF::FileSystemImpl::markPurgeable):
* Source/WTF/wtf/FileSystem.h:
* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/cocoa/FileSystemCocoa.mm:
(WTF::FileSystemImpl::markPurgeable):
* Source/WebCore/loader/appcache/ApplicationCacheStorage.h:
(WebCore::ApplicationCacheStorage::cacheDirectory const):
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::invalidateAndCancel):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/WebProcess/Model/mac/ARKitInlinePreviewModelPlayerMac.h:
* Source/WebKit/WebProcess/Model/mac/ARKitInlinePreviewModelPlayerMac.mm:
(WebKit::ARKitInlinePreviewModelPlayerMac::modelElementCacheDirectory):
(WebKit::modelElementCacheDirectory): Deleted.
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::didClose):
* Source/WebKit/WebProcess/WebProcess.h:

Canonical link: <a href="https://commits.webkit.org/257606@main">https://commits.webkit.org/257606@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ff5b8a54e19f50e10efc258f9585d5c5ac6a9ef0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99396 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8597 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32514 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108781 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169022 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103393 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9160 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85908 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/91925 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106704 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105153 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/6928 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90470 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33905 "layout-tests (failure)") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/88747 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21817 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/76790 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/90083 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2456 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23334 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/85934 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2369 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45727 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28894 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5230 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/8526 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42809 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/88801 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4242 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19862 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WTF/wtf/PlatformHave.h:
```diff
@@ -1355,3 +1355,7 @@
 #define HAVE_CONTINUITY_CAMEARA 1
 #endif
+
+#if PLATFORM(MAC) && !PLATFORM(IOS_SIMULATOR) && __has_include(<apfs/apfs_fsctl.h>)
```
Do we really need `!PLATFORM(IOS_SIMULATOR)` if we already have `PLATFORM(MAC)`?

cdumez replied:
Also, any reason we're not doing this on iOS as well?

szewai replied:
ah nice catch, the first condition should be `PLATFORM(COCOA)`

cdumez commented on Source/WTF/wtf/cocoa/FileSystemCocoa.mm:
```diff
@@ -239,6 +243,20 @@ bool setExcludedFromBackup(const String& path, bool excluded)
+    return fsctl(fileSystemPath.data(), APFSIOC_MARK_PURGEABLE, &flags, 0) != -1;
+#endif
+
+    return false;
```
Would be good for this to be in a `#else` case. Weird to have two return statements otherwise.

cdumez commented on Source/WebCore/loader/appcache/ApplicationCacheStorage.h:
```diff
@@ -56,6 +56,7 @@ class ApplicationCacheStorage : public RefCounted<ApplicationCacheStorage> {
     }
 
 
+    String cacheDirectory() const { return m_cacheDirectory; }
```
`const String&`

cdumez commented on Source/WebKit/WebProcess/WebProcess.h:
```diff
@@ -533,6 +533,7 @@ class WebProcess : public AuxiliaryProcess
     friend class WebConnectionToUIProcess;
     void didReceiveMessage(IPC::Connection&, IPC::Decoder&) override;
     bool didReceiveSyncMessage(IPC::Connection&, IPC::Decoder&, UniqueRef<IPC::Encoder>&) override;
+    void didClose(IPC::Connection&) override;
```
final?

philn replied:
So now `AuxiliaryProcess::didClose()` is no longer called. Leaving WebProcess running forever, in glib ports at least.

szewai replied:
Oh I saw the fix is landed in [257721@main](https://commits.webkit.org/257721@main); thanks for fixing!

cdumez commented on Source/WTF/wtf/cocoa/FileSystemCocoa.mm:
```diff
@@ -239,6 +243,20 @@ bool setExcludedFromBackup(const String& path, bool excluded)
+
+#if HAVE(APFS_CACHEDELETE_PURGEABLE)
+    uint64_t flags = APFS_MARK_PURGEABLE | APFS_PURGEABLE_DATA_TYPE | APFS_PURGEABLE_MARK_CHILDREN;
+    return fsctl(fileSystemPath.data(), APFSIOC_MARK_PURGEABLE, &flags, 0) != -1;
```
We should probably check `== 0`. I believe this function returns an error code so the value will differ depending on the error.

szewai replied:
I thought fsctl returns -1 for error but existing usage seems your suggestion is right; so I will change.

szewai committed:
Mark cache directories as purgeable
https://bugs.webkit.org/show_bug.cgi?id=247828
rdar://101927216

Reviewed by Chris Dumez.

These directories were purgeable by CacheDelete because they were under Caches/ directory by default. After 255271@main,
the directories of a WebsiteDataStore created with UUID can be outside of Caches/ directory, so we need to explicitly
mark them purgeable to keep the old behavior. Notice the files are unmarked if they are modified after being marked, so
we set the flag before process exits or session is destroyed.

* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::setExcludedFromBackup):
(WTF::FileSystemImpl::markPurgeable):
* Source/WTF/wtf/FileSystem.h:
* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/cocoa/FileSystemCocoa.mm:
(WTF::FileSystemImpl::markPurgeable):
* Source/WebCore/loader/appcache/ApplicationCacheStorage.h:
(WebCore::ApplicationCacheStorage::cacheDirectory const):
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::invalidateAndCancel):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/WebProcess/Model/mac/ARKitInlinePreviewModelPlayerMac.h:
* Source/WebKit/WebProcess/Model/mac/ARKitInlinePreviewModelPlayerMac.mm:
(WebKit::ARKitInlinePreviewModelPlayerMac::modelElementCacheDirectory):
(WebKit::modelElementCacheDirectory): Deleted.
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::didClose):
* Source/WebKit/WebProcess/WebProcess.h:

Canonical link: https://commits.webkit.org/257606@main
